File: src/main/java/alfheim/common/entity/boss/ai/AIChase.java
package alfheim.common.entity.boss.ai;

import java.util.Collections;

import alexsocol.asjlib.ASJUtilities;
import alfheim.common.entity.boss.EntityFlugel;
import net.minecraft.entity.player.EntityPlayer;
import vazkii.botania.common.core.helper.Vector3;

public class AIChase extends AIBase {

	public AIChase(EntityFlugel flugel, AITask task) {
		super(flugel, task);
	}

	@Override
	public void startExecuting() {
		int s = flugel.getStage();
		int i = s == 1 ? 300 : s == 2 ? 150 : 50;
		flugel.setAITaskTimer(flugel.worldObj.rand.nextInt(i) + i);
	}

	@Override
	public boolean continueExecuting() {
		flugel.checkCollision();
		if (flugel.getAITaskTimer() % 10 == 0) {
			String name = "";
			try {
				name = ASJUtilities.mapGetKeyOrDefault(flugel.playersWhoAttacked, Collections.max(flugel.playersWhoAttacked.values()), "");
			} catch (Throwable e) {
				e.printStackTrace();
				return canContinue();
			}
			
			EntityPlayer target = flugel.worldObj.getPlayerEntityByName(name);
		
			if (target != null) {
				Vector3 mot = new Vector3(target.posX - flugel.posX, target.posY - flugel.posY, target.posZ - flugel.posZ).normalize();
				flugel.motionX = mot.x;
				flugel.motionY = mot.y;
				flugel.motionZ = mot.z;
			} else {
				flugel.playersWhoAttacked.remove(name);
			}
		}
		
		return canContinue();
	}
}

File: src/main/java/alfheim/common/entity/boss/ai/AIChase.kt
package alfheim.common.entity.boss.ai;

import alexsocol.asjlib.ASJUtilities
import alfheim.common.entity.boss.EntityFlugel
import vazkii.botania.common.core.helper.Vector3
import java.util.Collections

class AIChase(flugel: EntityFlugel, task: AITask) : AIBase(flugel, task) {
    
    override fun startExecuting() {
        val s = flugel.getStage()
        val i = if (s == 1) 300 else if (s == 2) 150 else 50
        flugel.setAITaskTimer(flugel.worldObj.rand.nextInt(i) + i)
    }

    override fun continueExecuting(): Boolean {
        flugel.checkCollision()
        if (flugel.getAITaskTimer() % 10 == 0) {
            var name: String
            try {
                name = ASJUtilities.mapGetKeyOrDefault(flugel.playersWhoAttacked, Collections.max(flugel.playersWhoAttacked.values), "")
            } catch (e: Throwable) {
                e.printStackTrace()
                return canContinue()
            }
            val target = flugel.worldObj.getPlayerEntityByName(name)
            if (target != null) {
                val mot = Vector3(target.posX - flugel.posX, target.posY - flugel.posY, target.posZ - flugel.posZ).normalize()
                flugel.motionX = mot.x
                flugel.motionY = mot.y
                flugel.motionZ = mot.z
            } else {
                flugel.playersWhoAttacked.remove(name)
            }
        }
        return canContinue()
    }
}